xcm/src/v1/multi_asset.rs

```diff
@@ -0,0 +1,380 @@
+/// the desired meaning/effect. This means that relative paths should always by constructed to be read from the point of
+/// view of the receiving system, *which may be have a completely different meaning in the authoring system*.
+///
+/// Concrete identifiers are the preferred way of identifying an asset since they are entirely unambiguous.
+///
+/// A concrete identifier is represented by a `MultiLocation`. If a system has an unambiguous primary asset (such as
+/// Bitcoin with BTC or Ethereum with ETH), then it will conventionally be identified as the chain itself. Alternative
+/// and more specific ways of referring to an asset within a system include:
+///
+/// - `<chain>/PalletInstance(<id>)` for a Frame chain with a single-asset pallet instance (such as an instance of the
+///   Balances pallet).
+/// - `<chain>/PalletInstance(<id>)/GeneralIndex(<index>)` for a Frame chain with an indexed multi-asset pallet instance
+///   (such as an instance of the Assets pallet).
+/// - `<chain>/AccountId32` for an ERC-20-style single-asset smart-contract on a Frame-based contracts chain.
+/// - `<chain>/AccountKey20` for an ERC-20-style single-asset smart-contract on an Ethereum-like chain.
+///
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Encode, Decode, Debug)]
+pub enum MultiAsset {
+	/// No assets. Rarely used.
+	None,
+
+	/// All assets. Typically used for the subset of assets to be used for an `Order`, and in that context means
+	/// "all assets currently in holding".
+	All,
+
+	/// All fungible assets. Typically used for the subset of assets to be used for an `Order`, and in that context
+	/// means "all fungible assets currently in holding".
+	AllFungible,
+
+	/// All non-fungible assets. Typically used for the subset of assets to be used for an `Order`, and in that
+	/// context means "all non-fungible assets currently in holding".
+	AllNonFungible,
+
+	/// All fungible assets of a given abstract asset `id`entifier.
+	AllAbstractFungible { id: Vec<u8> },
+
+	/// All non-fungible assets of a given abstract asset `class`.
+	AllAbstractNonFungible { class: Vec<u8> },
+
+	/// All fungible assets of a given concrete asset `id`entifier.
+	AllConcreteFungible { id: MultiLocation },
+
+	/// All non-fungible assets of a given concrete asset `class`.
+	AllConcreteNonFungible { class: MultiLocation },
+
+	/// Some specific `amount` of the fungible asset identified by an abstract `id`.
+	AbstractFungible { id: Vec<u8>, #[codec(compact)] amount: u128 },
+
+	/// Some specific `instance` of the non-fungible asset whose `class` is identified abstractly.
+	AbstractNonFungible { class: Vec<u8>, instance: AssetInstance },
+
+	/// Some specific `amount` of the fungible asset identified by an concrete `id`.
+	ConcreteFungible { id: MultiLocation, #[codec(compact)] amount: u128 },
+
+	/// Some specific `instance` of the non-fungible asset whose `class` is identified concretely.
+	ConcreteNonFungible { class: MultiLocation, instance: AssetInstance },
+}
```

Comment:
The `All*` variants are filters, the other variants (e.g. `ConcreteFungible`) are (something like) identifiers.
See other comments for elaboration.